fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/explain/Attribute.java

```diff
@@ -21,42 +21,150 @@
+
+    /**
+     * Return the underlying object this attribute refers to.
+     * @return the object.
+     */
     @Nonnull
-    private final Object reference;
+    Object getReference();
+
+    /**
+     * Basic interface all attributes of {@link AbstractPlannerGraph.AbstractNode}
+     * as well as {@link AbstractPlannerGraph.AbstractEdge} that are use to serialize to gml.
+     */
+    interface GmlAttribute extends Attribute {
+        @Override
+        default <N, E> boolean isVisible(GraphExporter<N, E>.ExporterContext context) {
+            return context.getExporter() instanceof GmlExporter;
+        }
+    }
 
-    protected Attribute(final Object reference) {
-        this.reference = reference;
+    /**
+     * Interface for attributes of {@link AbstractPlannerGraph.AbstractNode}
+     * as well as {@link AbstractPlannerGraph.AbstractEdge} that are used to serialize to dot.
+     */
+    interface DotAttribute extends Attribute {
+        @Override
+        default <N, E> boolean isVisible(GraphExporter<N, E>.ExporterContext context) {
+            return context.getExporter() instanceof DotExporter;
+        }
     }
 
     /**
-     * Returns whether this method is semantic of it contains visual cues of other non-semantic information.
-     *
-     * @return {@code true} if the attribute is semantic, {@code false} otherwise.
+     * Interface for attributes of {@link AbstractPlannerGraph.AbstractNode}
+     * as well as {@link AbstractPlannerGraph.AbstractEdge} that are used to serialize to gml or to dot.
      */
-    public abstract boolean isSemanticAttribute();
+    interface CommonAttribute extends GmlAttribute, DotAttribute {
+        @Override
+        default <N, E> boolean isVisible(GraphExporter<N, E>.ExporterContext context) {
+            return true;
+        }
+    }
 
     /**
-     * Return the underlying object this attribute refers to.
-     * @return the object.
+     * Interface for attributes of {@link AbstractPlannerGraph.AbstractNode}
+     * as well as {@link AbstractPlannerGraph.AbstractEdge} that are used to neither serialize to gml or to dot.
+     * Attributes of this kind can only be referred to via variable substitution.
      */
-    @Nullable
-    public Object getReference() {
-        return reference;
+    interface InvisibleAttribute extends GmlAttribute, DotAttribute {
+        @Override
+        default <N, E> boolean isVisible(GraphExporter<N, E>.ExporterContext context) {
+            return false;
+        }
     }
 
-    @Override
-    public String toString() {
-        return Objects.requireNonNull(reference).toString();
+    /**
+     * Static factory method to create a GML attribute based on a reference to some object.
+     * @param reference the reference
+     * @return a new attribute that is only visible to the GML exporter
+     */
+    static GmlAttribute gml(@Nonnull final Object reference) {
+        return new GmlAttribute() {
```

Comment:
Is there a reason that these are all anonymous classes, rather than instantiations of some simple implementation class. If this sits on a hot code path, I think that the performance differences might matter?

Reply:
I would like to offer for users to potentially create their own sort of attributes but I would like to mix in the `gml()`, or the `dot()` flavor as well. It's nicer to provide interfaces if possible. The fact that they are anonymous classes does not make this implementation slower, the JVM just creates an implementation in-place with a cryptic name and uses that as a class name.